invoice/src/amount.rs

```diff
@@ -0,0 +1,303 @@
+    CentiMicro = 8,
+    Nano = 9,
+    DeciNano = 10,
+    CentiNano = 11,
+    Pico = 12,
+    DeciPico = 13,
+    CentiPico = 14,
+    Femto = 15,
+    DeciFemto = 16,
+    CentiFemto = 17,
+    Atto = 18,
+}
+impl StrictSerialize for Precision {}
+impl StrictDeserialize for Precision {}
+
+impl Precision {
+    pub fn from_strict_val_unchecked(value: &StrictVal) -> Self { value.unwrap_enum() }
+    pub const fn decimals(self) -> u8 { self as u8 }
+    pub const fn decimals_u32(self) -> u32 { self as u8 as u32 }
+
+    pub const fn multiplier(self) -> u64 {
+        match self {
+            Precision::Indivisible => 1,
+            Precision::Deci => 10,
+            Precision::Centi => 100,
+            Precision::Milli => 1000,
+            Precision::DeciMilli => 10_000,
+            Precision::CentiMilli => 100_000,
+            Precision::Micro => 1_000_000,
+            Precision::DeciMicro => 10_000_000,
+            Precision::CentiMicro => 100_000_000,
+            Precision::Nano => 1_000_000_000,
+            Precision::DeciNano => 10_000_000_000,
+            Precision::CentiNano => 100_000_000_000,
+            Precision::Pico => 1_000_000_000_000,
+            Precision::DeciPico => 10_000_000_000_000,
+            Precision::CentiPico => 100_000_000_000_000,
+            Precision::Femto => 1_000_000_000_000_000,
+            Precision::DeciFemto => 10_000_000_000_000_000,
+            Precision::CentiFemto => 100_000_000_000_000_000,
+            Precision::Atto => 1_000_000_000_000_000_000,
+        }
+    }
+
+    pub fn unchecked_convert(self, amount: impl Into<u64>) -> Amount {
+        (amount.into() * self.multiplier()).into()
+    }
+
+    pub fn checked_convert(self, amount: impl Into<u64>) -> Option<Amount> {
+        amount
+            .into()
+            .checked_mul(self.multiplier())
+            .map(Amount::from)
+    }
+    pub fn saturating_convert(self, amount: impl Into<u64>) -> Amount {
+        amount.into().saturating_mul(self.multiplier()).into()
+    }
+}
+
+impl From<Precision> for u16 {
+    fn from(value: Precision) -> Self { value as u8 as u16 }
+}
+
+impl From<Precision> for u32 {
+    fn from(value: Precision) -> Self { value as u8 as u32 }
+}
+
+impl From<Precision> for u64 {
+    fn from(value: Precision) -> Self { value as u8 as u64 }
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+pub struct CoinAmount {
+    pub int: u64,
+    pub fract: u64,
+    pub precision: Precision,
+}
+
+impl CoinAmount {
+    pub fn with(value: impl Into<Amount>, precision: impl Into<Precision>) -> Self {
+        let precision = precision.into();
+        let value = value.into();
+        let (int, fract) = value.split(precision);
+        CoinAmount {
+            int,
+            fract,
+            precision,
+        }
+    }
+}
+
+impl Display for CoinAmount {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let mut int = self.int.to_string();
+        if f.alternate() {
```

Comment:
Unsafe is required because of `from_utf8_unchecked` - since we are sure the string is UTF8

Chunks of three bytes to print the separators: `100'000'000`